core/src/main/scala/com/banno/kafka/consumer/ConsumerOps.scala

```diff
@@ -383,36 +419,60 @@ case class ConsumerOps[F[_], K, V](consumer: ConsumerApi[F, K, V]) {
           consumer.commitSync(offsets)
         val commitNextOffsets: F[Unit] =
           state.get.map(_.nextOffsets).flatMap(commit)
-        consumer
-          .recordStream(pollTimeout)
-          .evalMap { record =>
+        stream
+          .evalMap { a =>
             for {
-              a <- process(record)
-              s <- state.updateAndGet(_.update(record))
+              b <- process(a)
+              s <- state.updateAndGet(_.update(offsets(a), count(a)))
               now <- C.monotonic
               () <- s
                 .needToCommit(maxRecordCount, now, maxElapsedTime)
                 .traverse_(os =>
                   commit(os) *>
                   state.update(_.reset(now))
                 )
-            } yield a
+            } yield b
           }
           .onFinalize(commitNextOffsets)
       }
 
+  /*
+  I think this impl based on Stream.groupWithin works, but the Bs are output in groups, instead of as As are processed
+  Not sure if this matters or not... at the very least, some tests would need to be rewritten
+  If A is a batch of records, groupWithin also groups by number of batches, not total count of records
+   */
+  def processingAndCommitting2[A, B](
+      maxRecordCount: Int,
+      maxElapsedTime: FiniteDuration,
+  )(
+      stream: Stream[F, A],
+      process: A => F[B],
+      offsets: A => Map[TopicPartition, Long],
+  )(implicit F: Temporal[F]): Stream[F, B] =
+    stream
+      .evalMap(a => process(a).map(b => (offsets(a), b)))
+      .groupWithin(maxRecordCount, maxElapsedTime)
+      .evalTap { c =>
+        val offsets = c.foldLeft(Map.empty[TopicPartition, Long])(_ ++ _._1)
+        val nextOffsets =
+          offsets.view.mapValues(o => new OffsetAndMetadata(o + 1)).toMap
+        println(
```

Comment:
Should this `println` be removed, or replaced with proper logging?

Reply:
Definitely removed 😄 this is some experimental code to try out `Stream.groupWithin`. Will clean up before merge.